### DISCUSSION TITLE
Export-PnPTermGroupToXml from "site collection" term store

### DISCUSSION THREAD

**Discussion by mfxits** [pnp/powershell#2988] [Q&A]
Hi,

I'm using PnP.Powershell 1.12.0 and want to export a term group from the "site collection term store" of the root site collection.

Let's say the group is named "**ABC**". It's the only group in the "site local" terms store.

`Get-PnpTermGroup` lists all groups from the global term store including the local "ABC". 
For any global group piped to Export-PnPTermGroupToXml the result is as expected:  

`Get-PnpTermGroup -Identity "XYZ" | Export-PnPTermGroupToXml`

```
<pnp:TermGroups xmlns:pnp="http://schemas.dev.office.com/PnP/2022/09/ProvisioningSchema">
  <pnp:TermGroup Name="XYZ" ID="8574c038-e187-435d-b2e5-e0520ffafea7" Description="" UpdateBehavior="Overwrite">
    <pnp:TermSets> ... </pnp:TermSets> 
  </pnp:TermGroup>
</pnp:TermGroups>
```

But when I do the same with "ABC", the result is empty. It doesn't matter if I use the ID or name as parameter for "Identity".

`Get-PnpTermGroup -Identity "ABC" | Export-PnPTermGroupToXml`

```
   
```
Both groups returned from `Get-PnpTermGroup` have the type "Microsoft.SharePoint.Client.Taxonomy.TermGroup" and I can't see any difference between them.

I also played with `Get-PnPSiteCollectionTermStore`, but that command just returns the same term group object.

I searched for the topic but found nothing. Can anyone help please?

- Markus

**Comment by AndersRask** [comment]
Are your user term store admin? Could be a permission issue maybe?

fre. 7. apr. 2023 15.50 skrev mfxits ***@***.***>:

**Reply by mfxits** [reply]
Hi Anders,

thank you for your reply.

Yes, the user is term store admin and I used the same account to create the term groups in the term store management UI. 
Furthermore, the user can access the group's sets and terms using the other cmdlets like `Get-PnpTaxonomyItem` or `Get-PnpTermSet`.

I came across this "bug" on a customer's tenant and rebuilt the scenario on a test tenant. In both cases `Export-PnpTermGroupToXml` works for groups in the global store, but not for the site store (at least the root site collection store - I haven't tried another SC).

I need to export and re-import whole groups with all terms and custom properties. I thought the cmdlet would just do that, and writing my own script would be reinventing the wheel.

Could it  be a glitch only in version 1.12.0?

Markus

**Reply by AndersRask** [reply]
You say you can reproduce this, so I guess you have some XML that you import?

If so, check the xmlns:pnp for provisioningschema and see if they match.
I have seen XML extracted by older versions of PnP not be compatible with newer (or the other way around, I cant remember) so try and use the one that works for you.

**Reply by mfxits** [reply]
No, there's no XML to import _yet_, since `Export-PnpTermGroupToXml` doesn't produce it for the groups in the site local store. 

I could, of course, move all term sets in the local groups to new groups in the global store. But if I wanted to automate this I could also just write the im-/export script by myself. 

However, I will file this as bug and see what happens.

**Reply by AndersRask** [reply]
Yeah I can see what you mean now.

The cmdlet basically fetches a Site Template with TermGroups handlers, so you could fetch the XML that way too:

`Get-PnPSiteTemplate -Handlers TermGroups -IncludeAllTermGroups -IncludeSiteCollectionTermGroup `

When you parse the XML you will notice that local term groups are named {sitecollectiontermgroupname} and ID is set to 00000000-0000-0000-0000-000000000000, so the code fails to match on ID and name for that reason.

This is probably something that happens in the underlying API calls and outside the teams control.

Either use above site template cmdlet or fetch using 

`export-PnPTermGroupToXml -Identity "{sitecollectiontermgroupname}"`
or
`export-PnPTermGroupToXml -Identity 00000000-0000-0000-0000-000000000000`

**Reply by mfxits** [reply]
Thank you Anders.

Interestingly, a second group in the local store (here "SAP SIP Y") is reported correctly with all IDs set.

But not only is the first group's ID missing, which wouldn't hurt, but also all IDs of its term sets, too:

`export-PnPTermGroupToXml`

```
<pnp:TermGroup Name="{sitecollectiontermgroupname}" ID="00000000-0000-0000-0000-000000000000" Description="" SiteCollectionTermGroup="true" UpdateBehavior="Overwrite">
    <pnp:TermSets>
      <pnp:TermSet Name="SapSipX_1" ID="00000000-0000-0000-0000-000000000000" Description="">
        <pnp:CustomProperties>
          <pnp:Property Key="SapSipX_1_SCP1" Value="Kardamom" />
        </pnp:CustomProperties>
        <pnp:Terms>
          <pnp:Term Name="SapSipX_1_T1" ID="00000000-0000-0000-0000-000000000000" Owner="i:0#.f|membership|cod01@m365x38030475.onmicrosoft.com" Description="" SourceTermId="40549369-07d7-49a9-9ae8-3e08dff27c9d">
            <pnp:Terms />
            <pnp:CustomProperties>
              <pnp:Property Key="SapSipX_1_T1_SCP1" Value="Salz" />
            </pnp:CustomProperties>
            <pnp:LocalCustomProperties>
              <pnp:Property Key="SapSipX_1_T1_LCP1" Value="Pfeffer" />
            </pnp:LocalCustomProperties>
          </pnp:Term>
          <pnp:Term Name="SapSipX_1_T2" ID="00000000-0000-0000-0000-000000000000" Owner="i:0#.f|membership|cod01@m365x38030475.onmicrosoft.com" Description="" SourceTermId="b8ca95dd-5255-4d65-b239-85794771748a">
            <pnp:Terms />
          </pnp:Term>
        </pnp:Terms>
      </pnp:TermSet>
      <pnp:TermSet Name="SapSipX_2" ID="00000000-0000-0000-0000-000000000000" Description="">
        <pnp:Terms>
          <pnp:Term Name="SapSipX_2_T1" ID="00000000-0000-0000-0000-000000000000" Owner="i:0#.f|membership|cod01@m365x38030475.onmicrosoft.com" Description="" SourceTermId="c235fdaa-42b0-415e-a824-c190b672d30d">
            <pnp:Terms />
          </pnp:Term>
        </pnp:Terms>
      </pnp:TermSet>
    </pnp:TermSets>
  </pnp:TermGroup>
  <pnp:TermGroup Name="SAP SIP Y" ID="f1fc9668-0c45-40ad-8319-378aac183101" Description="" UpdateBehavior="Overwrite">
    <pnp:TermSets>
      <pnp:TermSet Name="SapSipY_1" ID="4dc33b39-678e-4a97-aabf-dc23bd369bec" Description="">
        <pnp:Terms>
          <pnp:Term Name="SapSipY_1_T1" ID="496c789a-a7df-4539-a4b6-58bb53d7c804" Owner="i:0#.f|membership|cod01@m365x38030475.onmicrosoft.com" Description="" SourceTermId="496c789a-a7df-4539-a4b6-58bb53d7c804">
            <pnp:Terms />
            <pnp:Labels>
              <pnp:Label Language="1033" Value="Balduin" />
              <pnp:Label Language="1033" Value="Egon" />
            </pnp:Labels>
          </pnp:Term>
        </pnp:Terms>
      </pnp:TermSet>
    </pnp:TermSets>
  </pnp:TermGroup>

```
Those missing term set IDs again are show stoppers, since I can't query their ID's without the containing group ID or name:

`Get-PnpTermSet -identity SapSipX_1 -TermGroup 00000000-0000-0000-0000-000000000000`

```
_Get-PnPTermSet: Object reference not set to an instance of an object._
```

For the time being I will try to convince my customer to only use global term sets...
